### PR TITLE
920 - compute and store the hash of smart contracts code when available

### LINKED ISSUE #920: Store a hash of smart contracts on origination
Use a conseiljs-compatible [1] hashing protocol to identify any originated smart contract by its content and refer to it on later operations (e.g. filtering, verification).

Allow naming of each hash to improve usability by clients and users.

[1] the reference code is [here](https://github.com/Cryptonomic/ConseilJS/blob/46988e4b388fd92ad88ff2bf036ab327fff980cb/src/chain/tezos/contracts/TezosContractUtils.ts#L34)

### PR DESCRIPTION
**Pull request by ivanopagano**
closes #920 

## required schema changes

We require to add an additional column to the `accounts` table, namely

```
    script_hash character varying
```

### CODE REVIEW

**Comment by piotrkosecki** [comment]
I've got one question - shouldn't this be added to the `accounts_history` too?

**Review by piotrkosecki — APPROVED** [review]
overall code looks ok, left one question.

**Comment by ivanopagano** [comment]
> I've got one question - shouldn't this be added to the `accounts_history` too?

😱

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    val blake2s = new Blake2s128()
+    blake2s.update(in.getBytes(StandardCharsets.UTF_8))
+    Hex.encode(blake2s.digest())
+  }
+
+  /** Remove unwanted characters and create a uniform layout for the contract parts,
+    * i.e. code, parameters, storage.
+    * Makes sure that contract hashing is consistent.
+    *
+    * @param code the michelson code as a raw string
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:47]: Nonexistent parameter.

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * We fold over the lines, keeping track of state with a partial result map and the current keyword.
+     * As we fold the keyword to line pair is added to the map using cats apis.
+     * The monoidal operator |+| appends together entries under the same key of two maps, and returns
+     * the resulting combined map, which is quite useful for our case.
+     */
+
+    /* checks for a keyword */
+    def startingKeyword(line: String) = keywords.find(sym => line.startsWith(sym.name))
+
+    def stripTralingComments(line: String) = line.replaceFirst("""\#[\s\S]+$""", "").trim()
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:61]: Do you need the `\#` in the regex?

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
+package tech.cryptonomic.conseil.indexer.tezos.michelson.contracts
+
+import java.nio.charset.StandardCharsets
+import scorex.util.encode.{Base16 => Hex}
+import org.bouncycastle.jcajce.provider.digest.Blake2s.Blake2s128
+import cats.data.Chain
+import cats.implicits._
+
+/** Generic utilities related to smart contracts on tezos */
+object SmartContracts {
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:10]: To follow our naming convention so far, can we call this `SmartContractUtils`?

**Review by vishakh — CHANGES_REQUESTED** [review]

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala (hunk 2 above)
Comment (ivanopagano) [reply conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:61]: it removes the comments, which are introduced by the `#` symbol

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala (hunk 3 above)
Comment (ivanopagano) [reply conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:10]: sure

Comment (ivanopagano) [reply conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:10]: First consider this: the code is inside a different package and module than the other *Utils.
The latter usually live in the `common` module under package `tech.cryptonomic.conseil.common.util`

This lives in package `tech.cryptonomic.conseil.indexer.tezos.michelson.contracts`
alongside
 - TokenContracts.scala
 - TNSContract.scala

Which in a sense makes the naming more consistent with the siblings

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala (hunk 1 above)
Comment (ivanopagano) [reply conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/michelson/contracts/SmartContracts.scala:47]: fixed

**Review by vishakh — APPROVED** [review]

**Pull request merged**